### PR TITLE
Remove JavaScript layer and rework API in UART module

### PR DESCRIPTION
**Pull request by LaszloLango**
Related issue: #1367

IoT.js-DCO-1.0-Signed-off-by: László Langó llango.u-szeged@partner.samsung.com

### CODE REVIEW

**Comment by hs0225** [comment]
@LaszloLango I cannot found the code for `on('data')`. Without this API, we have no way to read serial communications.

**Comment by LaszloLango** [comment]
@hs0225 The patch is ready for review. I've tested it on RPi2 and TizenRT.

**Comment by hs0225** [comment]
@LaszloLango Please add a test case for the `openSync`. What about changing the [line](https://github.com/LaszloLango/iotjs/blob/f28d09140a6a242c1e5674acc53da519d5a19a12/test/run_pass/test_uart.js#L35)?

**Comment by LaszloLango** [comment]
@hs0225 good idea. I've added a test for `openSync`.

File: src/modules/iotjs_module_uart.c [hunk 1]
```diff
@@ -199,159 +123,231 @@ static void iotjs_uart_after_worker(uv_work_t* work_req, int status) {
...
-  jerry_release_value(jdata_bits);
+  jerry_value_t jcallback = JS_GET_ARG_IF_EXIST(1, function);
 
-  UART_ASYNC(open, uart, jcallback, kUartOpOpen);
+  // If the callback doesn't exist, it is completed synchronously.
+  // Otherwise, it will be executed asynchronously.
+  if (!jerry_value_is_null(jcallback)) {
+    UART_CALL_ASYNC(kUartOpOpen, jcallback);
+  } else if (!iotjs_uart_open(uart)) {
+    return JS_CREATE_ERROR(COMMON, "UART Error: cannot open UART");
```
Comment (bzsolt) [inline src/modules/iotjs_module_uart.c]: I think we can use `uart_error_str()` here as well.

File: src/modules/iotjs_module_uart.c [hunk 2]
```diff
@@ -199,159 +123,231 @@ static void iotjs_uart_after_worker(uv_work_t* work_req, int status) {
...
-      return JS_CREATE_ERROR(COMMON, "UART Write Error");
-    }
+  uart->buf_data = JS_GET_ARG(0, string);
+  uart->buf_len = iotjs_string_size(&uart->buf_data);
+
+  bool result = iotjs_uart_write(uart);
+  iotjs_string_destroy(&uart->buf_data);
+
+  if (!result) {
+    return JS_CREATE_ERROR(COMMON, "UART Write Error");
```
Comment (bzsolt) [inline src/modules/iotjs_module_uart.c]: Ditto.

File: src/modules/iotjs_module_uart.c [hunk 3]
```diff
@@ -21,139 +21,78 @@
...
-}
-
-
-static iotjs_uart_t* iotjs_uart_instance_from_jval(jerry_value_t juart) {
-  iotjs_handlewrap_t* handlewrap = iotjs_handlewrap_from_jobject(juart);
-  return (iotjs_uart_t*)handlewrap;
-}
-
-
 iotjs_uart_reqwrap_t* iotjs_uart_reqwrap_from_request(uv_work_t* req) {
```
Comment (bzsolt) [inline src/modules/iotjs_module_uart.c]: This function can be static.

**Comment by LaszloLango** [comment]
@bzsolt thanks for the review. I've updated the PR.

**Comment by bzsolt** [comment]
LGTM (informally)

**Comment by yichoi** [comment]
LGTM

**Pull request merged**